### ISSUE TITLE
1.5.0 version is not adding transaction to block intermittently /reporting wrong block number

### ISSUE THREAD

**Issue by IamSandipC** [ethereum/go-ethereum#2498]
#### System information

Geth version: 1.5.0 
OS & Version: Windows
Commit hash : 

#### Expected behaviour
transaction added to the block mentioned as block number in transaction receipt 

#### Actual behaviour
Intermittently that is not the case. For some transactions (say one in 20), when transaction receipt is obtained, it has the block number. Doing web3.eth.getBlock('blockNumebr') returns empty transactions array.

#### Steps to reproduce the behaviour
1. Issue observed in a private blockchain mining almost at the rate 1 block every 1-2 seconds.
2. Added  series of similar transactions (same contract function call ->function writes an event to the tx log) to the blockchain at a rate say 1 per sec.
3. for some of the transactions ( say 1 in 20), after I get the transaction receipt,I take the block number.

Here is a txReceipt obtanes with web3.eth.getTransactionReceipt
```
Object {blockHash: "0xc722cff6d3b386e2d709cfb987f0f8e6587fbdd6466954edcdcc4607cddb54d1", blockNumber: 479, contractAddress: null, cumulativeGasUsed: 366186, from: "0x2e984ba2e1e51e44832c741de8fbd42d2cd3ad81"…}
blockHash
:
"0xc722cff6d3b386e2d709cfb987f0f8e6587fbdd6466954edcdcc4607cddb54d1"
blockNumber
:
479
contractAddress
:
null
cumulativeGasUsed
:
366186
from
:
"0x2e984ba2e1e51e44832c741de8fbd42d2cd3ad81"
gasUsed
:
183093
logs
:
Array[1]
root
:
"8199d8d46945d65c1a2541f6d33c66615fcdf885d227265243656178c97d47bc"
to
:
"0xd7e2e53f11edcf0d040440484084575f8f9e3c6d"
transactionHash
:
"0x91586eb2da6837b9e1b04b1c8ba594dfa0934adcfc09f69afec31e579bf42b31"
transactionIndex
:
1
__proto__
:
Object
```

4. I do a web3.eth.getBlock('blockNumber). It returns the block but without the transaction for which I got the receipt.

Here is the block with above mentioned block number
```
> web3.eth.getBlock(479)
{
  difficulty: 161681,
  extraData: "0xd983010500844765746887676f312e352e318777696e646f7773",
  gasLimit: 84079456,
  gasUsed: 0,
  hash: "0xcfa355f5c6ca55b9495a62daaf92f96e7a6d5b0b138b73c00902b562c5e68ca3",
  logsBloom: "0x0000000000000000000000000000000000000000000000000000000000000000
00000000000000000000000000000000000000000000000000000000000000000000000000000000
00000000000000000000000000000000000000000000000000000000000000000000000000000000
00000000000000000000000000000000000000000000000000000000000000000000000000000000
00000000000000000000000000000000000000000000000000000000000000000000000000000000
00000000000000000000000000000000000000000000000000000000000000000000000000000000
000000000000000000000000000000000000000000000000",
  miner: "0x2e984ba2e1e51e44832c741de8fbd42d2cd3ad81",
  nonce: "0x0c2a1519efcd5ee2",
  number: 479,
  parentHash: "0xf6c666bb5ddbb6a3562a7dcde8215844f623020909fd459f0a5daa6dfb02348
9",
  receiptRoot: "0x56e81f171bcc55a6ff8345e692c0f86e5b48e01b996cadc001622fb5e363b4
21",
  sha3Uncles: "0x1dcc4de8dec75d7aab85b567b6ccd41ad312451b948a7413f0a142fd40d4934
7",
  size: 540,
  stateRoot: "0x0de1b0ba91e241cb0aa6fa50eec6cf3ea497009b128e078887cd331e11030305
",
  timestamp: 1461782648,
  totalDifficulty: 69816343,
  transactions: [],
  transactionsRoot: "0x56e81f171bcc55a6ff8345e692c0f86e5b48e01b996cadc001622fb5e
363b421",
  uncles: []
}

```

#### Backtrace

````
[backtrace]
````

**Comment by karalabe** [comment]
Aren't more than one miners in your private network racing for blocks? So it can happen that a block is reorged by another one resulting in a new set of transaction receipts too. If you manage to query the receipt when the first block is imported but access the block when the second, it might explain the issue.

Instead of querying the block by number, please try querying the block by hash (that should also be included in the receipt). Btw, if you look at your example above, the block hashes are different too: the receipt contains `0xc722cff6d3b386e2d709cfb987f0f8e6587fbdd6466954edcdcc4607cddb54d1` as the block hash, whereas when you query the block, you get back `0xcfa355f5c6ca55b9495a62daaf92f96e7a6d5b0b138b73c00902b562c5e68ca3`. So the original block which fired the receipt was replaced.

**Comment by obscuren** [comment]
Please make sure that "removed" isn't set on the receipt.

**Comment by IamSandipC** [comment]
@karalabe No. It is just a development node.. a single node which is my machine. No other node is in the network... 

But yes... you made a good point. In a proper network it has to be the hash not the number. Thank you for educating :)

But interesting why the multiple forks are created for a single node mining network.

**Comment by IamSandipC** [comment]
@obscuren how to check the "removed" in the receipt? Is it a prop in json obj?
@karalabe - one more thing.. If block is reorged and I do a fresh call for getTxRcpt, it should give me the new correct block number and hash. right? In my case that is not happening. It is still pointing to the old (invalid) block number and hash.
